File: interviews/leetcode/leetcode-75/binary-tree-dfs/1448. Count Good Nodes in Binary Tree/solution_recursive.go
package main

import "fmt"

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// Helper function for DFS traversal
func dfs(node *TreeNode, maxVal int) int {
	if node == nil {
		return 0
	}

	count := 0
	if node.Val >= maxVal {
		count = 1
	}

	// Update the max value and recurse for left and right children
	maxVal = max(maxVal, node.Val)
	count += dfs(node.Left, maxVal)
	count += dfs(node.Right, maxVal)

	return count
}

func goodNodes(root *TreeNode) int {
	return dfs(root, root.Val)
}

// Helper function to create a binary tree from a slice
func createBinaryTree(lst []interface{}) *TreeNode {
	if len(lst) == 0 || lst[0] == nil {
		return nil
	}
	root := &TreeNode{Val: lst[0].(int)}
	queue := []*TreeNode{root}
	i := 1
	for i < len(lst) {
		node := queue[0]
		queue = queue[1:]
		if lst[i] != nil {
			node.Left = &TreeNode{Val: lst[i].(int)}
			queue = append(queue, node.Left)
		}
		i++
		if i < len(lst) && lst[i] != nil {
			node.Right = &TreeNode{Val: lst[i].(int)}
			queue = append(queue, node.Right)
		}
		i++
	}
	return root
}

func main() {
	root := createBinaryTree([]interface{}{3, 1, 4, 3, nil, 1, 5})
	result := goodNodes(root)
	fmt.Println(result) // Output: 4
}

File: interviews/leetcode/leetcode-75/binary-tree-dfs/1448. Count Good Nodes in Binary Tree/solution_recursive.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:

        # Start DFS traversal with the root node
        return dfs(root, root.val)

# Helper function for DFS traversal
def dfs(node, max_val):
    if not node:
        return 0
    # If the current node is a good node, increment the count
    count = 1 if node.val >= max_val else 0
    # Update the max value encountered on the path
    max_val = max(max_val, node.val)
    # Recurse for left and right children
    count += dfs(node.left, max_val)
    count += dfs(node.right, max_val)
    return count


# Helper function to create a binary tree from a list (for testing purposes)
def create_binary_tree(lst):
    if not lst:
        return None
    root = TreeNode(lst[0])
    queue = [root]
    i = 1
    while i < len(lst):
        node = queue.pop(0)
        if lst[i] is not None:
            node.left = TreeNode(lst[i])
            queue.append(node.left)
        i += 1
        if i < len(lst) and lst[i] is not None:
            node.right = TreeNode(lst[i])
            queue.append(node.right)
        i += 1
    return root

# Test the solution
# if __name__ == "__main__":
#     solution = Solution()
#     root = create_binary_tree([3, 1, 4, 3, None, 1, 5])
#     result = solution.goodNodes(root)
#     print(result)  # Output: 4
